### ISSUE TITLE
1.8.16: deduction guides generate warnings

### ISSUE THREAD

**Issue by skypjack** [doxygen/doxygen#7451]
I've just updated doxygen locally to version 1.8.16 and it started to generate random (wrong) warnings for deduction guides (C++17). Doxygen 1.8.15 worked as expected instead.

Here is an example from [a project of mine](https://github.com/skypjack/entt):
```
/**
 * @brief Deduction guide.
 *
 * It allows to deduce the character type of the hashed string directly from a
 * human-readable identifer provided to the constructor.
 *
 * @tparam Char Character type.
 * @tparam N Number of characters of the identifier.
 * @param str Human-readable identifer.
 */
template<typename Char, std::size_t N>
basic_hashed_string(const Char (&str)[N]) ENTT_NOEXCEPT
-> basic_hashed_string<Char>;
```

The error I obtain is definitely wrong since there is no return type to document here:
>warning: return type of member entt::basic_hashed_string is not documented

The error is reproducible. Just clone the repository and compile the tests (from within `build` run `cmake .. && make test`).
It's not the only new and wrong error actually. The same project is also a good test case for issue #7212 if you want.

**Comment by albert-github** [comment]
With the mentioned snippet and a standard Doxyfile I was not able to represent the problem.
Running against the given repository is, probably, a bit overkill and furthermore is is a moving target so:
- Can you please attach a, small, self contained example (source+configuration file in a tar or zip) that allows us to reproduce the problem?  Please don't add external links as they might not be persistent.

**Comment by skypjack** [comment]
I tried to track this down but I failed to create a minimal, reproducible example.
Afaik it's actually due to the coexistence of a function-like macro used in a _forward-file_ (a file with forward declarations for the classes of the library) and a deduction guide in a file that includes the previous one.
I've spent almost two hours trying to reproduce it with a small example but that library is all I have to reproduce it at the moment, I'm sorry.

It's interesting because I thought it was due to the deduction guide itself. Instead, that's only where it shows up.
If I replace the macro call with the actual code, the problem disappears.

It looks **a lot** like a quite extreme corner case that doxygen cannot manage properly for some unknown reasons (I'm not into the code to put any idea on the table unfortunately).

**Comment by skypjack** [comment]
Ok, I think I've found the key to reproduce it.
I'm leaving now but I'll try to provide you with a minimal example as soon as possible.

**Comment by skypjack** [comment]
Here is the code required to reproduce the error:
```
namespace bar {
 
/*! @brief my class */
template<typename>
struct foo {  
    /**
     * @brief constructor
     * @tparam U tparam
     */
    template<typename U>
    foo(U) {}  
};
 
/**
 * @brief guide    
 * @tparam U tparam
 */
template<typename U>
foo(U) -> foo<U *>;
 
}
 
/**
 * @namespace bar
 * @brief my namespace
 */
```
Just generate a config file with `doxygen -g` and set `WARN_NO_PARAMDOC` to `YES`.
Note that the error disappears if you remove the documentation for the namespace.

**Comment by skypjack** [comment]
As a side note, I've the same error for function-like macros at namespace scope. Everything worked as expected with v1.8.15 as above.
I strongly suspect the problem is the same though. Should I open a new issue for that or can we track it here?

**Comment by albert-github** [comment]
It is hard to tell whether or not it is the same issue. A test case won't hurt, so I think best is to create a new issue (so it can be handled separately in case it are different problems) with also a reference to this issue.

**Comment by skypjack** [comment]
Ok, see #7456 (shameless plug to have an explicit cross reference). :+1:

**Comment by doxygen** [comment]
Look like the same issue indeed. See this [commit](https://github.com/doxygen/doxygen/commit/7cf6e640540954d974374b28f692e4d05ef22049) for a possible solution.

**Comment by skypjack** [comment]
@doxygen afaik if the test above passes, it's right. I'm not quite into the code to review it otherwise, I'm sorry.

**Comment by doxygen** [comment]
This issue was previously marked 'fixed but not released',
which means it should be fixed in doxygen version 1.8.17.
Please verify if this is indeed the case. Reopen the
issue if you think it is not fixed and please include any additional information
that you think can be relevant (preferably in the form of a self-contained example).
